File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.cc
#include<iostream>
#include<queue>
using namespace std;

int N, K;
bool visited[100001] = {}; // 방문했는지 확인

void bfs(int n) {
	queue<pair<int,int>> q;
	q.push(make_pair(n,0));
	visited[n] = true;

	while (!q.empty()) {
		n = q.front().first;
		int count = q.front().second;
		q.pop();
		if (n == K) {
			cout << count << "\n";
			break;
		}

		if (n - 1 >= 0 && n - 1 <= 100000) {
			if (!visited[n - 1]) {
				q.push(make_pair(n - 1, count + 1));
				visited[n - 1] = true;
			}
		}

		if (n + 1 >= 0 && n + 1 <= 100000) {
			if (!visited[n + 1]) {
				q.push(make_pair(n + 1, count + 1));
				visited[n + 1] = true;
			}
		}

		if (n*2 >= 0 && n*2 <= 100000) {
			if (!visited[n*2]) {
				q.push(make_pair(n*2, count + 1));
				visited[n*2] = true;
			}
		}
	}
}


int main() {
	cin >> N >> K;
	bfs(N);
}

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.util.*;
import java.io.*;
public class Main{
    static boolean[] visited;
    static int bfs(int now, int target , int count){
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{now,count});
        visited[now] = true;
        while(!q.isEmpty()){
            int[] current = q.poll();
            if(current[0] == target) return current[1];
            if(current[0]>target) {
                if (current[0] - 1 <=100000 && current[0] - 1 >=0 && !visited[current[0] - 1]) {
                    q.offer(new int[]{current[0] - 1, current[1] + 1});
                    visited[current[0] - 1] = true;
                }
            }
            else{
                if (current[0] *2 <=100000 && current[0] *2>=0 && !visited[current[0]*2]) {
                    q.offer(new int[]{current[0] *2, current[1] + 1});
                    visited[current[0] *2] = true;
                }
                if (current[0] + 1 <=100000 && current[0]+1>=0 && !visited[current[0] +1]) {
                    q.offer(new int[]{current[0] +1, current[1] + 1});
                    visited[current[0] + 1] = true;
                }
                if (current[0] - 1 <=100000 && current[0]-1>=0 && !visited[current[0] - 1]) {
                    q.offer(new int[]{current[0] - 1, current[1] + 1});
                    visited[current[0] - 1] = true;
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        visited = new boolean[100001];

        System.out.println(bfs(N,K,0));
    }
}
